### ISSUE TITLE
Bug: [unbound-method] Does not fail for destructured parameters

### ISSUE THREAD

**Issue by steverep** [typescript-eslint/typescript-eslint#8947]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&fileType=.ts&code=MYGwhgzhAEAKD2ECWAXJ8B20DeBYAUNNAMQAe0AvNAAwDcBRxAnpTffg9KQBQCUO0TkQBOAUxQBXYVhQALJBAB0ZdkQC%2BBTkz44h0MZOnQ5C5U1XQN%2BKwQBmEjMDSZot%2BPG4AHAIwAuOIio6BgANAKk-qQATGFM-kxRlv4IyM4Y-HiE0MCYECjhkd6x8d6WrD4WBlJYANqk3joA1FxRfLEN-M0JfAC67FZAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=&tokens=false

### Repro Code

```TypeScript
class Position {
  #x = 0;
  #y = 0;

  x() { 
    return this.#x;
  }

  y() {
    return this.#y;
  }
}

function foo(p1: Position, { x: x2, y: y2 }: Position) {
  const { x: x1, y: y1 } = p1;
  return [x1() + x2(), y1() + y2()];
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/unbound-method": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expect the rule to fail when destructuring `x2` and `y2` in the second function parameter.

### Actual Result

There are errors when destructuring `x1` and `y1` in the function body as expected, but no error for the equivalent code within the function parameter.

**Comment by kirkwaiblinger** [comment]
To be clear, this is what you want to happen, right?:
```ts
type HasMethod = { method(): void };
// actual
function f({ method }: HasMethod) {
}
// expected
function f({ method }: HasMethod) {
             ^^^^^^
}
```

**Comment by steverep** [comment]
Yes, that destructured assignment should trigger the rule the same way as if it were destructured in the body.

**Comment by kirkwaiblinger** [comment]
Awesome, thanks for the report! Seems like a bug to me as well.

Looking at past work it looks like this was probably just a simple oversight when the code was added that handles other cases of destructuring (https://github.com/typescript-eslint/typescript-eslint/pull/2228). Might be as simple as adding `ObjectPattern` to the rule.

**Comment by targos** [comment]
I understand that it can be seen as a bug fix, but this is probably going to break many consumers who use the rule. For example, in my project, this is now flagged:

```tsx
import { expect, test } from '@playwright/experimental-ct-react';

import SomeComponent from './SomeComponent';

test('some component', async ({ mount }) => {
  const result = await mount(<SomeComponent />);
  // ...
});
```

Note that Playwright makes it mandatory to use destructuring here so the only escape hatch is to disable the rule in tests.
______

It also affects React components with props defined like this:

```tsx
interface TestProps {
  renderSomething(): string;
}

export function Test(props: TestProps) {
  const { renderSomething } = props;
  return <div>{renderSomething()}</div>
}
```

In that second case, I could rewrite my interface to fix it:

```ts
interface TestProps {
  renderSomething: () => string;
}
```

**Comment by JoshuaKGoldberg** [comment]
> Note that Playwright makes it mandatory to use destructuring

Do you have a link for that?

> In that second case, I could rewrite my interface to fix it:

Indeed, that would be a more correct way to write that type. +1 that this is a _bug fix_. The `(): ...` method declaration was incorrect to begin with.

**Comment by targos** [comment]
> > Note that Playwright makes it mandatory to use destructuring
> 
> Do you have a link for that?
> 

I can't find a direct mention in the documentation, but this is part of the [test fixtures](https://playwright.dev/docs/test-fixtures) feature.

[This is where](https://github.com/microsoft/playwright/blob/8412d973c03ecfce9c0dbabe0e3fec6d307cdc2d/packages/playwright/src/common/fixtures.ts#L279-L282) they check for it.

**Comment by kirkwaiblinger** [comment]
I'm with @JoshuaKGoldberg - the change is a correct bug fix. So this issue is correctly closed as completed.

@targos Thanks for bringing this pain point up, though! Would you kindly spin off a new issue to continue the conversation? I think it merits discussion, so I wouldn't want this conversation to get buried by virtue of occurring in a closed issue.
